### PR TITLE
Add config for custom login page

### PR DESCRIPTION
**Pull request by crazyrohila**
Added an extra config parameter to redirect to the custom login page when CDSSO is enabled

### CODE REVIEW

File: src/policyagent/policy-agent-options.ts [hunk 1]
```diff
@@ -43,6 +43,11 @@ export interface PolicyAgentOptions {
    */
   appUrl?: string;
 
+  /**
+   * The web agent Id to check for conditional URLs
+   */
+  webAgentId?: string;
+
```
Comment (niroz89) [inline src/policyagent/policy-agent-options.ts]: may not needed, since agent profile username and webAgentId are same.

**Review by niroz89 — COMMENTED** [review]
Tested the PR, it was throwing 401 authentication error.

I suppose below is the cause for it.

https://github.com/SPHTech/node-openam-agent/blob/e2f63a3c2e5c4ab47550b1c18a18d401d133d89b/src/policyagent/policy-agent.ts#L129

In the above code agent authenticate doesn't pass any module. It should be as below,
`.authenticate(username, password, realm, '', 'Application')`

Tested locally with the above change, It throws below error for the `GET /amserver/json/realm-config/agents/WebAgent/{webAgentId}?realm=%2F` endpoint.

` data: {
      code: 403,
      reason: 'Forbidden',
      message: 'The user has insufficient privileges'
    }`

Checked the session details endpoint, it gives below properties but not the custom properties. 

```
"properties": {
        "Locale": "en_US",
        "authInstant": "2019-11-02T04:49:55Z",
        "Organization": "o=example.com.sg",
        "UserProfile": "Required",
        "Principals": "demo-node-agent-1",
        "successURL": "https://app.example.com/abc/goto.html",
        "CharSet": "UTF-8",
        "Host": "127.0.0.1",
        "FullLoginURL": "/amserver/UI/Login?authIndexValue=Application&authIndexType=Module&realm=%2F",
        "PostAuthProcessInstance": "com.example.AMPostAuthProcess",
        "AuthLevel": "0",
        "clientType": "genericHTML",
        "AMCtxId": "85053bef-5551-4de5-80c1-c5406e9f8730-387095",
        "loginURL": "/amserver/UI/Login",
        "UserId": "demo-node-agent-1",
        "AuthType": "Application",
        "IndexType": "module_instance",
        "sun.am.UniversalIdentifier": "id=demo-node-agent-1,ou=agent,o=example.com.sg",
        "HostName": "127.0.0.1",
        "amlbcookie": "02",
        "Principal": "id=demo-node-agent-1,ou=agent,o=example.com.sg",
        "UserToken": "demo-node-agent-1"
    }
```

**Comment by niroz89** [comment]
@crazyrohila @riteshgurung We should be using below endpoint to get the custom login properties.
```
curl -X GET https://am.example.com/amserver/json/realms/root/agents/demo-node-agent-1 \
-H 'Cookie: dsiPlanetDirectoryPro=g99KALzbAl3p4DmTwDIDpVLgJzk.*AAJTSQACMDUAAlNLABxUb0ZvM29ROWRUZDk1cUpYSTZBc2lZSkpCVDA9AAR0eXBlAANDVFMAAlMxAAIwMg..*'
```
Conditional login url is available in `"com.sun.identity.agents.config.freeformproperties"` field.
```
{
    "_id": "demo-node-agent-1",
    "username": "demo-node-agent-1",
    "realm": "/",
     .....
     "com.sun.identity.agents.config.freeformproperties": [
        "com.forgerock.agents.conditional.login.url[0]=app.example.com:8000|https://ds-acc-auth.sphdigital.com/SPHAuth/login?svc=st_online",
        "AgentType=WebAgent",
        "org.forgerock.openam.agents.config.allow.custom.login=true",
        "org.forgerock.agents.config.cdsso.original.url.redirect.param=mycustomgoto",
        "org.forgerock.agents.config.cdsso.txcookie.maxage=300"
    ]
    ....
}
```

**Comment by crazyrohila** [comment]
@niroz89 I have got the properties from this endpoint; removed the agentID param from config.

File: src/policyagent/policy-agent.ts [hunk 2]
```diff
@@ -433,10 +433,20 @@ export class PolicyAgent extends EventEmitter {
...
+    // store queryparams first if login url has any
+    const extratQueryParams = conditionalUrlKey.split('?')[1];
+    // Keep without query parameters part
+    conditionalUrlKey = conditionalUrlKey.split('?')[0];
+    // Get conditional url values with conditions
     const conditionalUrlVal = conditionalUrlKey.split("=")[1];
+    // Split condition with actual login url
     if (conditionalUrlVal && conditionalUrlVal.indexOf("|") > -1) {
       if (this.options.appUrl.indexOf(conditionalUrlVal.split("|")[0]) > -1) {
         loginUrl = conditionalUrlVal.split("|")[1];
```
Comment (niroz89) [inline src/policyagent/policy-agent.ts]: @crazyrohila , com.forgerock.agents.conditional.login.url is multi valued property, there can be com.forgerock.agents.conditional.login.url[1], com.forgerock.agents.conditional.login.url[2] etc, correct way should be we need to get the map of domain & login url and use the correct one according to the application domain. 

`com.forgerock.agents.conditional.login.url[0]=app.example.com:8000|https://ds-acc-auth.sphdigital.com/SPHAuth/login?svc=st_online`

Comment (crazyrohila) [reply src/policyagent/policy-agent.ts]: Aha, my bad. I missed this one, pushed the changes to get a map of `condition => URL` and then checked for the current app in passed in config.

**Comment by niroz89** [comment]
Tested ACK.

**Pull request merged**